chain/network/src/raw.rs

```diff
@@ -343,3 +267,233 @@ impl Peer {
+            match m {
+                raw::ReceivedMessage::AnnounceAccounts(_) => Self::AnnounceAccounts,
+                raw::ReceivedMessage::Pong { nonce, .. } => Self::Pong(nonce),
+            }
+        }
+    }
+
+    #[allow(dead_code)]
+    enum RecvResult {
+        Ok(&'static [ReceivedMessageType]),
+        ConnectErr,
+    }
+
+    #[allow(dead_code)]
+    struct TestCase {
+        msg_sequence: &'static [PeerMsgType],
+        want: RecvResult,
+    }
+
+    #[allow(dead_code)]
+    static TESTCASES: &[TestCase] = &[
+        TestCase {
+            msg_sequence: &[PeerMsgType::Handshake, PeerMsgType::Pong(1)],
+            want: RecvResult::Ok(&[ReceivedMessageType::Pong(1)]),
+        },
+        // no handshake = bad
+        TestCase {
+            msg_sequence: &[PeerMsgType::PeersRequest, PeerMsgType::Pong(1)],
+            want: RecvResult::ConnectErr,
+        },
+        // same as first one but w/ some ignored ones before
+        TestCase {
+            msg_sequence: &[
+                PeerMsgType::Handshake,
+                PeerMsgType::PeersRequest,
+                PeerMsgType::PeersRequest,
+                PeerMsgType::Pong(1),
+            ],
+            want: RecvResult::Ok(&[ReceivedMessageType::Pong(1)]),
+        },
+        TestCase {
+            msg_sequence: &[
+                PeerMsgType::Handshake,
+                PeerMsgType::PeersRequest,
+                PeerMsgType::Pong(1),
+                PeerMsgType::PeersRequest,
+                PeerMsgType::Pong(2),
+                PeerMsgType::PeersRequest,
+                PeerMsgType::Pong(3),
+            ],
+            want: RecvResult::Ok(&[
+                ReceivedMessageType::Pong(1),
+                ReceivedMessageType::Pong(2),
+                ReceivedMessageType::Pong(3),
+            ]),
+        },
+    ];
+
+    #[allow(dead_code)]
+    struct Server {
+        secret_key: SecretKey,
+        peer_id: PeerId,
+        listener: TcpListener,
+    }
+
+    #[allow(dead_code)]
+    async fn write_message<S: AsyncWrite + Unpin>(
+        stream: &mut S,
+        msg: &PeerMessage,
+    ) -> io::Result<()> {
+        let mut msg = msg.serialize(Encoding::Proto);
+        let mut buf = (msg.len() as u32).to_le_bytes().to_vec();
+        buf.append(&mut msg);
+        stream.write_all(&buf).await
+    }
+
+    #[allow(dead_code)]
+    impl Server {
+        async fn new() -> Self {
+            let secret_key = SecretKey::from_random(KeyType::ED25519);
+            let peer_id = PeerId::new(secret_key.public_key());
+            let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
+            Self { secret_key, peer_id, listener }
+        }
+
+        async fn run(self, msg_sequence: &'static [PeerMsgType]) -> anyhow::Result<()> {
+            let (mut stream, _addr) = self.listener.accept().await.unwrap();
+            let (read, mut write) = stream.split();
+            let mut read = tokio::io::BufReader::new(read);
+
```

Comment:
For your implementation we are interested in testing that it can communicate with the actual implementation of the PeerManager. Your current test implementation will not notice if some substantial change to the network protocol happened with requires changes to your ping tool. Please use crate::peer_manager::testonly::start to create a real instance of the PeerManager.

Reply:
true, done